consensus/polybft/bridge_event_relayer.go

```diff
@@ -0,0 +1,460 @@
+	"strings"
+
+	"github.com/0xPolygon/polygon-edge/consensus/polybft/contractsapi"
+	"github.com/0xPolygon/polygon-edge/consensus/polybft/wallet"
+	"github.com/0xPolygon/polygon-edge/contracts"
+	"github.com/0xPolygon/polygon-edge/crypto"
+	"github.com/0xPolygon/polygon-edge/txrelayer"
+	"github.com/0xPolygon/polygon-edge/types"
+	"github.com/Ethernal-Tech/blockchain-event-tracker/store"
+	"github.com/Ethernal-Tech/blockchain-event-tracker/tracker"
+	"github.com/Ethernal-Tech/ethgo"
+	"github.com/hashicorp/go-hclog"
+	bolt "go.etcd.io/bbolt"
+)
+
+var (
+	errUnknownBridgeEventRelayerEvent = errors.New("unknown event from gateway contract")
+	bridgeMessageResultEventSig       = new(contractsapi.BridgeMessageResultEvent).Sig()
+	eventChBuffer                     = 100
+)
+
+// BridgeEventRelayer is an interface that defines functions for bridge event relayer
+type BridgeEventRelayer interface {
+	EventSubscriber
+	AddLog(chainID *big.Int, eventLog *ethgo.Log) error
+	Close()
+	Start(runtimeCfg *runtimeConfig, eventProvider *EventProvider) error
+}
+
+var _ BridgeEventRelayer = (*dummyBridgeEventRelayer)(nil)
+
+// dummyBridgeEventRelayer is a dummy implementation of a BridgeEventRelayer
+type dummyBridgeEventRelayer struct{}
+
+func (d *dummyBridgeEventRelayer) AddLog(chainID *big.Int, eventLog *ethgo.Log) error { return nil }
+func (d *dummyBridgeEventRelayer) GetLogFilters() map[types.Address][]types.Hash {
+	return make(map[types.Address][]types.Hash)
+}
+func (d *dummyBridgeEventRelayer) ProcessLog(header *types.Header, log *ethgo.Log, dbTx *bolt.Tx) error {
+	return nil
+}
+func (d *dummyBridgeEventRelayer) Close() {}
+func (d *dummyBridgeEventRelayer) Start(runtimeCfg *runtimeConfig, eventProvider *EventProvider) error {
+	return nil
+}
+
+var _ BridgeEventRelayer = (*bridgeEventRelayerImpl)(nil)
+
+type bridgeEventRelayerImpl struct {
+	key    crypto.Key
+	logger hclog.Logger
+
+	state *BridgeMessageStore
+
+	externalTxRelayers map[uint64]txrelayer.TxRelayer
+	internalTxRelayer  txrelayer.TxRelayer
+	internalChainID    *big.Int
+
+	blockchain blockchainBackend
+
+	bridgeConfig  map[uint64]*BridgeConfig
+	eventTrackers []*tracker.EventTracker
+
+	eventCh chan contractsapi.StructAbi
+	quitCh  chan struct{}
+}
+
+// newBridgeEventRelayer creates a new instance of bridge event relayer
+// if the node is not a relayer, it will return a dummy bridge event relayer
+func newBridgeEventRelayer(
+	runtimeConfig *runtimeConfig,
+	logger hclog.Logger,
+) (BridgeEventRelayer, error) {
+	if !runtimeConfig.consensusConfig.IsRelayer {
+		return &dummyBridgeEventRelayer{}, nil
+	}
+
+	relayer := &bridgeEventRelayerImpl{
+		key:             wallet.NewEcdsaSigner(runtimeConfig.Key),
+		logger:          logger.Named("bridge-relayer"),
+		internalChainID: big.NewInt(runtimeConfig.genesisParams.ChainID),
+		blockchain:      runtimeConfig.blockchain,
+		eventCh:         make(chan contractsapi.StructAbi, eventChBuffer),
+		quitCh:          make(chan struct{}),
```

Comment:
I was thinking about that, but we are not passing context anywhere. It would be a big refactor. Maybe a task for polishing code after all this is done.